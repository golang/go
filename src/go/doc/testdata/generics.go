// Copyright 2021 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Package generics contains the new syntax supporting generic programming in
// Go.
package generics

// Variables with an instantiated type should be shown.
var X Type[int]

// Parameterized types should be shown.
type Type[P any] struct {
	Field P
}

// Constructors for parameterized types should be shown.
func Constructor[lowerCase any]() Type[lowerCase] {
	return Type[lowerCase]{}
}

// MethodA uses a different name for its receiver type parameter.
func (t Type[A]) MethodA(p A) {}

// MethodB has a blank receiver type parameter.
func (t Type[_]) MethodB() {}

// MethodC has a lower-case receiver type parameter.
func (t Type[c]) MethodC() {}

// Constraint is a constraint interface with two type parameters.
type Constraint[P, Q interface{ string | ~int | Type[int] }] interface {
	~int | ~byte | Type[string]
	M() P
}

// int16 shadows the predeclared type int16.
type int16 int

// NewEmbeddings demonstrates how we filter the new embedded elements.
type NewEmbeddings interface {
	string // should not be filtered
	int16
	struct{ f int }
	~struct{ f int }
	*struct{ f int }
	struct{ f int } | ~struct{ f int }
}

// Func has an instantiated constraint.
func Func[T Constraint[string, Type[int]]]() {}

// AnotherFunc has an implicit constraint interface.
//
// Neither type parameters nor regular parameters should be filtered.
func AnotherFunc[T ~struct{ f int }](_ struct{ f int }) {}

// AFuncType demonstrates filtering of parameters and type parameters. Here we
// don't filter type parameters (to be consistent with function declarations),
// but DO filter the RHS.
type AFuncType[T ~struct{ f int }] func(_ struct{ f int })

// See issue #49477: type parameters should not be interpreted as named types
// for the purpose of determining whether a function is a factory function.

// Slice is not a factory function.
func Slice[T any]() []T {
	return nil
}

// Single is not a factory function.
func Single[T any]() *T {
	return nil
}
