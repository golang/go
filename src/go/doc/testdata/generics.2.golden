// Package generics contains the new syntax supporting generic ...
PACKAGE generics

IMPORTPATH
	testdata/generics

FILENAMES
	testdata/generics.go

FUNCTIONS
	// AnotherFunc has an implicit constraint interface.  Neither type ...
	func AnotherFunc[T ~struct{ f int }](_ struct{ f int })

	// Func has an instantiated constraint. 
	func Func[T Constraint[string, Type[int]]]()

	// Single is not a factory function. 
	func Single[T any]() *T

	// Slice is not a factory function. 
	func Slice[T any]() []T


TYPES
	// AFuncType demonstrates filtering of parameters and type ...
	type AFuncType[T ~struct{ f int }] func(_ struct {
		// contains filtered or unexported fields
	})

	// Constraint is a constraint interface with two type parameters. 
	type Constraint[P, Q interface{ string | ~int | Type[int] }] interface {
		~int | ~byte | Type[string]
		M() P
	}

	// NewEmbeddings demonstrates how we filter the new embedded ...
	type NewEmbeddings interface {
		string	// should not be filtered
	
		struct {
			// contains filtered or unexported fields
		}
		~struct {
			// contains filtered or unexported fields
		}
		*struct {
			// contains filtered or unexported fields
		}
		struct {
			// contains filtered or unexported fields
		} | ~struct {
			// contains filtered or unexported fields
		}
		// contains filtered or unexported methods
	}

	// Parameterized types should be shown. 
	type Type[P any] struct {
		Field P
	}

	// Variables with an instantiated type should be shown. 
	var X Type[int]

	// Constructors for parameterized types should be shown. 
	func Constructor[lowerCase any]() Type[lowerCase]

	// MethodA uses a different name for its receiver type parameter. 
	func (t Type[A]) MethodA(p A)

	// MethodB has a blank receiver type parameter. 
	func (t Type[_]) MethodB()

	// MethodC has a lower-case receiver type parameter. 
	func (t Type[c]) MethodC()

